csi/node_server.go

```diff
@@ -224,6 +230,17 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	return &csi.NodePublishVolumeResponse{}, nil
 }
 
+func (ns *NodeServer) collectWorkloadPodsStatus(volume *longhornclient.Volume, log *logrus.Entry) map[corev1.PodPhase][]string {
+	podsStatus := map[corev1.PodPhase][]string{}
+
+	for _, workload := range volume.KubernetesStatus.WorkloadsStatus {
+		phase := corev1.PodPhase(workload.PodStatus)
+		podsStatus[phase] = append(podsStatus[phase], workload.PodName)
+	}
+
+	return podsStatus
+}
```

Comment:
Do you test the RWX volume with multiple pods (part of them are running)?

Reply:
The `NodePublishVolume` should only proceed with the target mounting when the pod is being initiated (`Pending` state), AFAIK, the volume attachment doesn't occur in any other state. So this condition should also apply to RWX volume workloads.

For confirmation, I've tested the following scenario:
1. Deploy [rwx-nginx-deployment.yaml](https://github.com/longhorn/longhorn/blob/master/examples/rwx/rwx-nginx-deployment.yaml).
1. Wait until all workload pods are in the running state.
1. Delete one of the workload pods.
1. Wait for the deleted workload pod to restart and reach the running state.

Reply:
I mean, if this fix the handle the below RWX case:
1. Have a RWX volume with multiple running pods 
2. Kubelet on one node (volume attached node, or a normal node containing one running pod) is stopped for a while (in or out of the pod eviction time)
3. Kubelet get back and try to recover the RWX volume

Reply:
1. Create [rwx-nginx-deployment.yaml](https://github.com/longhorn/longhorn/blob/master/examples/rwx/rwx-nginx-deployment.yaml)
    ```bash
    curl https://raw.githubusercontent.com/longhorn/longhorn/master/examples/rwx/rwx-nginx-deployment.yaml | kubectl apply -f -
    ```
1. Wait for deployments running 
    ```bash
    > k get deployments.apps 
    NAME       READY   UP-TO-DATE   AVAILABLE   AGE
    rwx-test   3/3     3            3           36s
  
    > k -n longhorn-system get volume
    NAME                                       DATA ENGINE   STATE      ROBUSTNESS   SCHEDULED   SIZE         NODE            AGE
    pvc-8893d93f-f4dd-43ac-ade1-035a08b647a6   v1            attached   healthy                  1073741824   ip-10-0-2-121   40s
    ```
1. Stop kubelet for the volume node for 360 seconds
    ```bash
    > ssh ec2-user@10.0.2.121
    > systemctl stop k3s-agent && sleep 360 && systemctl start k3s-agent
    ```
1. Check deployment pods running
    ```bash
    > k get pod
    NAME                        READY   STATUS    RESTARTS   AGE
    rwx-test-79d5494fd8-9wpdc   2/2     Running   0          10m
    rwx-test-79d5494fd8-vpfsz   2/2     Running   0          10m
    rwx-test-79d5494fd8-zhvqd   2/2     Running   0          72s
    ```

Reply:
Is there any reason that you are concerned that this change might not work for RWX volume and its workloads?